go-controller/pkg/node/vrfmanager/vrf_manager.go

```diff
@@ -0,0 +1,309 @@
+	return nil
+}
+
+func (vrfm *Controller) reconcile() error {
+	vrfm.mu.Lock()
+	defer vrfm.mu.Unlock()
+	start := time.Now()
+	defer func() {
+		klog.V(5).Infof("Vrf Manager: reconciling VRFs took %v", time.Since(start))
+	}()
+	vrfsToKeep := make(map[string]*vrf)
+	for _, vrf := range vrfm.vrfs {
+		if vrf.delete {
+			err := vrfm.deleteVRF(vrf)
+			if err != nil {
+				return err
+			}
+			continue
+		}
+		err := vrfm.sync(vrf)
+		if err != nil {
+			return err
+		}
+		vrfsToKeep[vrf.name] = vrf
+	}
+	vrfm.vrfs = vrfsToKeep
+	return nil
+}
+
+func (vrfm *Controller) syncLink(link netlink.Link) error {
+	vrfm.mu.Lock()
+	defer vrfm.mu.Unlock()
+	vrf, ok := vrfm.vrfs[link.Attrs().Name]
+	if !ok {
+		return nil
+	}
+	return vrfm.sync(vrf)
+}
+
+func (vrfm *Controller) sync(vrf *vrf) error {
+	if vrf == nil {
+		return nil
+	}
+	vrfLink, err := util.GetNetLinkOps().LinkByName(vrf.name)
+	if err == nil {
+		if vrfLink.Type() != "vrf" {
+			return errors.New("node has another non vrf device with same name")
+		}
+		vrfDev, ok := vrfLink.(*netlink.Vrf)
+		if ok && vrfDev.Table != vrf.table {
+			return errors.New("found conflict with existing vrf device table id")
+		}
+	}
+	// Create Vrf device if it doesn't exist.
+	if util.GetNetLinkOps().IsLinkNotFoundError(err) {
+		vrfLink = &netlink.Vrf{
+			LinkAttrs: netlink.LinkAttrs{Name: vrf.name},
+			Table:     vrf.table,
+		}
+		err = util.GetNetLinkOps().LinkAdd(vrfLink)
+	}
+	if err != nil {
+		return err
+	}
+	vrfLink, err = util.GetNetLinkOps().LinkByName(vrf.name)
+	if err != nil {
+		return err
+	}
+	if vrfLink.Attrs().OperState != netlink.OperUp {
+		if err = util.GetNetLinkOps().LinkSetUp(vrfLink); err != nil {
+			return err
+		}
+	}
+	// Ensure enslave interface is synced up with the cache.
+	var existingEnslave string
+	links, err := util.GetNetLinkOps().LinkList()
+	if err != nil {
+		return err
+	}
+	for _, link := range links {
+		if link.Attrs().MasterIndex == vrfLink.Attrs().Index {
+			existingEnslave = link.Attrs().Name
+		}
+	}
+	if existingEnslave == vrf.enslaveInterface {
+		return nil
+	}
+	if existingEnslave != "" {
+		err = removeInterfaceFromVRF(vrf.name, existingEnslave)
+		if err != nil {
+			return fmt.Errorf("failed to remove inteface %s from vrf device: %s, err: %v", existingEnslave, vrf.name, err)
+		}
+	}
+	if vrf.enslaveInterface != "" {
+		err = enslaveInterfaceToVRF(vrf.name, vrf.enslaveInterface)
+		if err != nil {
+			return fmt.Errorf("failed to enslave inteface %s into vrf device: %s, err: %v", vrf.enslaveInterface, vrf.name, err)
+		}
+	}
+	vrfm.vrfs[vrf.name] = vrf
+	return nil
+}
+
+// AddVrf adds a Vrf device into the node.
+func (vrfm *Controller) AddVrf(name, enslaveInterface string, table uint32) error {
+	vrfm.mu.Lock()
+	defer vrfm.mu.Unlock()
+
+	_, ok := vrfm.vrfs[name]
+	if ok {
+		klog.V(5).Infof("Vrf Manager: vrf %s already found in the cache", name)
+		return nil
+	}
+	return vrfm.sync(&vrf{name, table, enslaveInterface, false})
```

Comment:
i think we should use whats in the cache or if it doesnt exist, create new vrf type, add to cache and call sync with it.